# IMDB Project

My aim of this project is to practice the data cleaning / basic analysis skills I have learnt from the Dataquest online learning platform.

I have downloaded an IMDB top 1000 movies dataset from Kaggle to use in my first independant project. I have  dirtied the dataset slightly so that I can implement a few of the techniques I have learnt. The data consists of 1000 of the highest rated movies over a ten year period (2006 - 2016).

## Data Dictionary

- Rank (#Column removed)
- TitleOfMovie
- GenreOfMovie - A list of geners seperated by commas
- DescriptionOfMovie
- DirectorOfMovie
- Actors - A list of actors seperated by commas
- Year
- Runtime (Minutes)
- Rating
- Votes
- Revenue (Millions)
- Metascore

### Metascore (Further information)
The metascore is a the weighted average review score from the worlds most well respected critics. 

Movies will get a Metascore only if at least four critics's reviews are collected.

The higher the Metascore, the more positive reviews a movie has.

[Link to Source of information](https://www.imdb.com/list/ls051211184/)

### Rating (Further information)
Everyday user of imdb can cast a vote between 1 and 10 for each movie. The rating for each movie is the average rating from voters that are not movie critics.


## Importing Pandas & NumPy Libraries & Reading Text File Into the Dataframe

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Reading the text file into a dataframe, assigning to variable movies
# an error ocurred using default utf-8 encoding had to be set to latin-1
movies = pd.read_csv('imdb-movie-data.csv',encoding='latin-1')

## Basic Data Exploration & Modifications

In [3]:
# Print the first five rows for the dataset
movies.head()

TitleOfMovie              GenreOfMovie  \
0  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1               Prometheus  Adventure,Mystery,Sci-Fi   
2                    Split           Horror,Thriller   
3                     Sing   Animation,Comedy,Family   
4            Suicide Squad  Action,Adventure,Fantasy   

                                  DescriptionOfMovie       DirectorOfMovie  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating        Votes  Revenue (Millions)  Metascore  
0     8.1   757,074.00              333,13       76.0  
1     7.0   485,820.00              126,46       65.0  
2     7.3   157,606.00              138,12       62.0  
3     7.2    60,545.00              270,32       59.0  
4     6.2   393,727.00              325,02       40.0

In [4]:
# View information about the data frame
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 11 columns):
TitleOfMovie          1000 non-null object
GenreOfMovie          1000 non-null object
DescriptionOfMovie    1000 non-null object
DirectorOfMovie       1000 non-null object
Actors                1000 non-null object
Year                  1000 non-null int64
Runtime (Minutes)     1000 non-null int64
Rating                1000 non-null float64
 Votes                1000 non-null object
Revenue (Millions)    872 non-null object
Metascore             936 non-null float64
dtypes: float64(2), int64(2), object(7)
memory usage: 86.0+ KB


We can see that the dataset consists of 1000 rows and eleven columns, the majority of which consist of non-null objects which are string values the remaining columns consist of numeric data such as integers and floating point numbers. We can also see that:-

- Column names need to be changed to camel_case.
- Revenue column currently a string value. Change to float.
- Votes column currently a string value.
- the Revenue and Metascore columns are missing data.


### Modifying Column Names

After reviewing the first five rows of the dataset we can see that the column names are currently in SnakeCase format rather than pythons preferred camel_case format. I will modify this using the dataframe.column attribute.

In [5]:
# I can view the column names of the dataset using the dataframe.columns attribute
movies.columns

Index(['TitleOfMovie', 'GenreOfMovie', 'DescriptionOfMovie', 'DirectorOfMovie',
       'Actors', 'Year', 'Runtime (Minutes)', 'Rating', ' Votes ',
       'Revenue (Millions)', 'Metascore'],
      dtype='object')

In [6]:
# Copy and paste column names / rename / assign back to dataframe.columns attribute
movies.columns = ['movie', 'genre', 'description', 'director',
       'actors', 'year', 'runtime_(minutes)', 'rating', 'votes',
       'revenue_(millions)', 'metascore']

In [7]:
movies.head(0)

Empty DataFrame
Columns: [movie, genre, description, director, actors, year, runtime_(minutes), rating, votes, revenue_(millions), metascore]
Index: []

### Convert Revenue Column to a Float
The values within the revenue column currently contain commas which will be replaced by full stops (period). The column will then be converted a floating point number.

In [8]:
# Modification to revenue column using .str.replace() and the .astype() methods
movies['revenue_(millions)'] = (movies['revenue_(millions)']
                                        .str.replace(',','.')
                                        .astype(float)
                               )


In [9]:
# Modification to votes column using .str.replace() and the .astype() methods
movies['votes'] = (movies['votes']
                           .str.replace(',','')
                           .str.replace('.00','')
                           .astype(int)
                  )

### View Descriptive Statistics of the Dataset

In [10]:
# Describe method to view numerical data only. set parameter include='all' for categorical data
movies.describe()

year  runtime_(minutes)       rating         votes  \
count  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean   2012.783000         113.172000     6.723200  1.635150e+05   
std       3.205962          18.810908     0.945429  1.893601e+05   
min    2006.000000          66.000000     1.900000  4.000000e+00   
25%    2010.000000         100.000000     6.200000  2.990775e+04   
50%    2014.000000         111.000000     6.800000  1.028255e+05   
75%    2016.000000         123.000000     7.400000  2.323410e+05   
max    2016.000000         191.000000     9.000000  1.791916e+06   

       revenue_(millions)   metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

We can see from the information above that the votes column above has 1000 values, a mean value of 163,000 votes, a minimum value of 400 votes and a max value of 1,791,916 votes.

I would like to find out how many rows of data within the votes column consist of such low values, as these values could question the integrity of the dataset. It is my understanding that if the average rating of a single movie is made up of a small amount of votes compared to others then it would be unreasonable to conclude that a movie with a metascore of 100 but with only 400 votes would be accurate. (REWORD BEFORE YOU UPLOAD TO GITHUB)

In [11]:
# Boolean selection of the movie and votes column for every movie with less than 1000 votes
less_than_1000_votes = movies[['movie','votes']][movies['votes'] < 10000]
less_than_1000_votes.describe()

votes
count   175.000000
mean   2554.737143
std    2860.387870
min       4.000000
25%     240.500000
50%    1176.000000
75%    4699.000000
max    9993.000000

We can see that the dataframe consists of 93 values with vote values that are less than 10000. These movies will be removed from the dataset.

In [12]:
movies = movies[movies['votes'].between(10000,180000000)]
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 825 entries, 0 to 999
Data columns (total 11 columns):
movie                 825 non-null object
genre                 825 non-null object
description           825 non-null object
director              825 non-null object
actors                825 non-null object
year                  825 non-null int64
runtime_(minutes)     825 non-null int64
rating                825 non-null float64
votes                 825 non-null int64
revenue_(millions)    786 non-null float64
metascore             785 non-null float64
dtypes: float64(3), int64(3), object(5)
memory usage: 77.3+ KB


We can determine from the information above that there are now 825 entries within our dataset however the revenue_(millions) and metascore contain missing values. 

- Metascore: All rows with missing data will be dropped.
- revenue_(millions): assign each missing value with the mean revenue.

In [18]:
# Confirm quantity of nan balues within metascore column
metascore_is_null = movies['metascore'].isnull()


False    785
True      40
Name: metascore, dtype: int64


The metascore column contains 40 nan values.

In [26]:
# Dropping the rows which contain nan values
movies = movies.dropna(subset=['metascore'])
# Checking the data with nan values has been dropped
movies['metascore'].isnull().value_counts()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 785 entries, 0 to 999
Data columns (total 11 columns):
movie                 785 non-null object
genre                 785 non-null object
description           785 non-null object
director              785 non-null object
actors                785 non-null object
year                  785 non-null int64
runtime_(minutes)     785 non-null int64
rating                785 non-null float64
votes                 785 non-null int64
revenue_(millions)    756 non-null float64
metascore             785 non-null float64
dtypes: float64(3), int64(3), object(5)
memory usage: 73.6+ KB


In [54]:
revenue_nan_values = movies[movies['revenue_(millions)'].isnull()]
revenue_nan_values.info()
#revenue_bool['revenue_(millions)'] = movies['revenue_(millions)'].mean()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29 entries, 61 to 995
Data columns (total 11 columns):
movie                 29 non-null object
genre                 29 non-null object
description           29 non-null object
director              29 non-null object
actors                29 non-null object
year                  29 non-null int64
runtime_(minutes)     29 non-null int64
rating                29 non-null float64
votes                 29 non-null int64
revenue_(millions)    0 non-null float64
metascore             29 non-null float64
dtypes: float64(3), int64(3), object(5)
memory usage: 2.7+ KB


In [58]:
# Replace all nan values within the revenue_(millions) column with the average revenue of the dataset.
movies['revenue_(millions)'].fillna(movies['revenue_(millions)'].mean(), inplace=True)
movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 785 entries, 0 to 999
Data columns (total 11 columns):
movie                 785 non-null object
genre                 785 non-null object
description           785 non-null object
director              785 non-null object
actors                785 non-null object
year                  785 non-null int64
runtime_(minutes)     785 non-null int64
rating                785 non-null float64
votes                 785 non-null int64
revenue_(millions)    785 non-null float64
metascore             785 non-null float64
dtypes: float64(3), int64(3), object(5)
memory usage: 73.6+ KB


In [59]:
# Sorting the top 30 movies within the dataset by the metascore (The Critics Review).
# Method chaining used to improve code readability
(movies[['movie', 'rating', 'votes', 'metascore']]
         .sort_values(by='metascore', ascending=False)
         .head(30)
)

movie  rating   votes  \
656                                          Boyhood     7.9  286722   
41                                         Moonlight     7.5  135095   
230                                  Pan's Labyrinth     8.2  498879   
509                                          Gravity     7.8  622089   
489                                      Ratatouille     8.0  504039   
111                                 12 Years a Slave     8.1  486338   
21                             Manchester by the Sea     7.9  134213   
501                                            Carol     7.2   77995   
406                                 Zero Dark Thirty     7.4  226661   
672                                  The Hurt Locker     7.6  352023   
241                                       Inside Out     8.2  416689   
192                                        Spotlight     8.1  268282   
611                                     Toni Erdmann     7.6   24387   
6                                         La La Land     8.3  258682   
688                                      Toy Story 3     8.3  586669   
136                           No Country for Old Men     8.1  660286   
173                                              Her     8.0  390531   
207                                         Paterson     7.5   26089   
347                                  American Hustle     7.3  379088   
67                                Mad Max: Fury Road     8.1  632842   
476                              The Lives of Others     8.5  278103   
413                                             Elle     7.2   35417   
988                                          Martyrs     7.1   63785   
685                                The King's Speech     8.0  534388   
260  Birdman or (The Unexpected Virtue of Ignorance)     7.8  440299   
238                         The Grand Budapest Hotel     8.1  530881   
499                                               Up     8.3  722203   
105                               Hell or High Water     7.7  115546   
311                                  La vie d'AdՌle     7.8  103150   
133                                         Whiplash     8.5  477276   

     metascore  
656      100.0  
41        99.0  
230       98.0  
509       96.0  
489       96.0  
111       96.0  
21        96.0  
501       95.0  
406       95.0  
672       94.0  
241       94.0  
192       93.0  
611       93.0  
6         93.0  
688       92.0  
136       91.0  
173       90.0  
207       90.0  
347       90.0  
67        90.0  
476       89.0  
413       89.0  
988       89.0  
685       88.0  
260       88.0  
238       88.0  
499       88.0  
105       88.0  
311       88.0  
133       88.0